src/Interpreters/Streaming/Aggregator.cpp

```diff
@@ -1566,59 +1445,60 @@ bool Aggregator::checkLimits(size_t result_size, bool & no_more_keys) const
 
 
 template <typename Method, typename Table>
-void Aggregator::convertToBlockImpl(
-    Method & method,
-    Table & data,
-    MutableColumns & key_columns,
-    AggregateColumnsData & aggregate_columns,
-    MutableColumns & final_aggregate_columns,
-    Arena * arena,
-    bool final,
-    bool clear_states) const
+Block Aggregator::convertToBlockImpl(
+    Method & method, Table & data, Arena * arena, Arenas & aggregates_pools, bool final, size_t rows, bool clear_states, AggregateStateType type) const
 {
     if (data.empty())
-        return;
-
-    if (key_columns.size() != params.keys_size)
-        throw Exception{"Aggregate. Unexpected key columns size.", ErrorCodes::LOGICAL_ERROR};
+    {
+        auto && out_cols = prepareOutputBlockColumns(params, aggregate_functions, getHeader(final), aggregates_pools, final, rows);
```

Comment:
Personally, i still like the `previous prepareBlockAndFill*` method since it is very clear : prepare, fill and finalize in one go, and caller pass in the fill callback.  Is there a strong reason we need need break it up to 3 functions ?

Reply:
`prepareBlockAndFill` Is old implementation. for now  community is new implementation and have some optimization on it. 
What I was thinking at the time was to make these changes to the new implementation to avoid doing it again in the future.

Reply:
This make sense. And let's cherry pick more regarding the aggregator improvements.